admission_control/admission-control-service/src/upstream_proxy.rs

```diff
@@ -20,184 +21,180 @@ use network::validator_network::{
-            role: config.get_role(),
+            role,
             client,
         }
     }
+}
 
-    /// main routine. starts sync coordinator that listens for CoordinatorMsg
-    pub async fn process_network_messages(
-        mut self,
-        network_events: Vec<AdmissionControlNetworkEvents>,
-    ) {
-        let mut events = select_all(network_events).fuse();
+/// main routine. starts sync coordinator that listens for CoordinatorMsg
+pub async fn process_network_messages(
+    upstream_proxy_data: UpstreamProxyData,
+    network_events: Vec<AdmissionControlNetworkEvents>,
+    executor: TaskExecutor,
+    mut client_events: mpsc::UnboundedReceiver<(
+        SubmitTransactionRequest,
+        oneshot::Sender<failure::Result<SubmitTransactionResponse>>,
+    )>,
+) {
+    let mut events = select_all(network_events).fuse();
+    let mut peer_info = upstream_proxy_data.peer_info.clone();
+    let workers_available = upstream_proxy_data.ac_config.max_concurrent_inbound_syncs;
+    let bounded_executor = BoundedExecutor::new(workers_available, executor);
 
-        loop {
-            ::futures::select! {
-                (mut msg, mut callback) = self.client_events.select_next_some() => {
-                    let result = self.submit_transaction_upstream(msg).await;
-                    if let Err(e) = callback.send(result) {
-                        error!("[admission control] failed to send back transaction result with error: {:?}", e);
-                    }
-                },
-                network_event = events.select_next_some() => {
-                    match network_event {
-                        Ok(event) => {
-                            match event {
-                                Event::NewPeer(peer_id) => {
-                                    debug!("[admission control] new peer {}", peer_id);
-                                    self.new_peer(peer_id);
-                                }
-                                Event::LostPeer(peer_id) => {
-                                    debug!("[admission control] lost peer {}", peer_id);
-                                    self.lost_peer(peer_id);
-                                }
-                                Event::RpcRequest((peer_id, mut message, callback)) => {
-                                    if let Some(AdmissionControlMsg_oneof::SubmitTransactionRequest(request)) = message.message {
-                                        if let Err(err) = self.process_submit_transaction_request(request, callback).await {
-                                            error!("[admission control] failed to process transaction request, peer: {}, error: {:?}", peer_id, err);
-                                        }
-                                    }
+    loop {
+        ::futures::select! {
+            (mut msg, mut callback) = client_events.select_next_some() => {
+                let result = submit_transaction_upstream(msg, &upstream_proxy_data.clone()).await;
```

Comment:
this should also be executed in other thread

Reply:
Not sure if this is the best approach but I think I need to create two different submit_transaction_upstream in that case because the callbacks are different in the case of intially sending a request and a full node sending a request